File: app/src/main/java/in/nerd_is/android_showcase/zhihu_daily/model/Story.java
package in.nerd_is.android_showcase.zhihu_daily.model;

import android.os.Parcelable;

import com.google.auto.value.AutoValue;
import com.ryanharter.auto.value.parcel.ParcelAdapter;
import com.squareup.moshi.JsonAdapter;
import com.squareup.moshi.Moshi;

import org.threeten.bp.LocalDate;

import in.nerd_is.android_showcase.common.lib_support.auto.LocalDateParcelAdapter;
import in.nerd_is.android_showcase.common.lib_support.sqldelight.LocalDateAdapter;
import in.nerd_is.android_showcase.common.lib_support.sqldelight.UrlListAdapter;

/**
 * @author Xuqiang ZHENG on 2016/10/7.
 */
@AutoValue
public abstract class Story implements StoryModel, Parcelable {

    private LocalDate date;

    @Override @ParcelAdapter(LocalDateParcelAdapter.class)
    public LocalDate date() {
        return date;
    }

    public void setDate(LocalDate date) {
        this.date = date;
    }

    // for sqldelight
    private static final LocalDateAdapter DATE_ADAPTER = new LocalDateAdapter();
    private static final UrlListAdapter LIST_ADAPTER = new UrlListAdapter();

    @SuppressWarnings("StaticInitializerReferencesSubClass")
    public static final Factory<Story> FACTORY = new Factory<>(
            (id, type, title, image, images, read, date) -> {
                AutoValue_Story story = new AutoValue_Story(id, type, title, image, images, read);
                story.setDate(date);
                return story;
            }, LIST_ADAPTER, DATE_ADAPTER);

    public static JsonAdapter<Story> jsonAdapter(Moshi moshi) {
        return new AutoValue_Story.MoshiJsonAdapter(moshi);
    }
}

File: app/src/main/kotlin/in/nerd_is/android_showcase/zhihu_daily/model/Story.kt
/*
 *    Copyright 2017 Xuqiang ZHENG
 *
 *    Licensed under the Apache License, Version 2.0 (the "License");
 *    you may not use this file except in compliance with the License.
 *    You may obtain a copy of the License at
 *
 *        http://www.apache.org/licenses/LICENSE-2.0
 *
 *    Unless required by applicable law or agreed to in writing, software
 *    distributed under the License is distributed on an "AS IS" BASIS,
 *    WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 *    See the License for the specific language governing permissions and
 *    limitations under the License.
 */

package `in`.nerd_is.android_showcase.zhihu_daily.model

import android.arch.persistence.room.Entity
import android.arch.persistence.room.PrimaryKey
import android.os.Parcel
import android.os.Parcelable
import org.threeten.bp.LocalDate
import paperparcel.PaperParcel

/**
 * @author Xuqiang ZHENG on 2017/6/12.
 */
@PaperParcel
@Entity
data class Story(@PrimaryKey val id: Long,
                 val type: Int = 0,
                 val title: String,
                 val image: String?,
                 val images: List<String>?,
                 val read: Boolean = false) : Parcelable {

    var date: LocalDate? = null

    companion object {
        @JvmField val CREATOR = PaperParcelStory.CREATOR
    }

    override fun writeToParcel(dest: Parcel, flags: Int) {
        PaperParcelStory.writeToParcel(this, dest, flags)
    }

    override fun describeContents() = 0
}